Split the tests into their own files

bencord0 opened the pull request:
Definitely looking for feedback on this one.

So far, I've just split them out into files with the same name as the functions, but we should rename or refactor them to better express what areas they help to test.

Ideas?

clux commented on tests/testmain.rs:
```diff
@@ -1,78 +1,22 @@
+    export_check, fetch_release_build_and_publish, get_environment_and_fetch, has_config_and_manifest,
+    init_force, list_everything, no_publish_non_release_builds, query_check, remove_dependencies,
+    run_scripts, setup, shell_echo, shell_permissions, status_on_experimentals, update_save, verify_checks,
+};
```
is globbing not more sensible here?

bencord0 replied:
There might be a better way to do this in "2018" edition, but I've been holding off thinking too much about the module layouts (and extern crate) until upgrading.

clux commented on tests/common/query_check.rs:
```diff
@@ -0,0 +1,7 @@
+
+    let rl = lal::query(backend, Some(env_name), "hello", true);
+    assert!(rl.is_ok(), "could query latest for hello");
+}
```
a lot of tiny files with very specific names after this refactor.

maybe we should make some of these functions match the subcommand we're testing.
e.g. this file contains all the tests for query?

bencord0 replied:
Yes. That's the plan.

Any thoughts about how to do this?
Or where to make the splits?

The problem I'm having is that most commands depend on other commands. (e.g. build probably depends on fetch, everything depends on configure).

bencord0 replied:
Okay, I've think I've got this now.

Each command gets it's own `test/cases/test_*.rs` file.
resolving this comment.

clux commented on tests/common/mod.rs:
```diff
@@ -0,0 +1,132 @@
+    demo_config.push("./configs/demo.json");
+
+    let mut testdir = PathBuf::from(env::var("LAL_CODE_DIR").unwrap());
+    testdir.push("tests");
```
we set the evar to unwrap read from it again? are we even picking up on the evar from tests anyway?

bencord0 replied:
Yes! Since I've removed all calls to `set_current_dir`, keeping a reference to the code isn't necessary any more.

clux commented on tests/testmain.rs:
```diff
@@ -1,78 +1,22 @@
-    copy(&from, state.tempdir.path(), &copy_options).expect("copy component to tempdir");
-    return to;
-}
+mod common;
```
don't you feel this feels a bit awkward to have the logic for `setup` inside the module above then run them all the same in this outer file? a lot of this here strikes me a bit as a custom test harness

bencord0 replied:
I tried `rstest`, which has fixtures but that was incompatible with `serial`. I might try it again.

`speculate` has a `before()` syntax similar to rspec.
I haven't tried that yet, but it looks like it could work too.

Know of any others?

bencord0 replied:
I've now moved all of the checks to their own (command specific) files. `tests/testmain.rs` now only has four lines in it, and isn't a wrapper for other tests anymore.

clux reviewed (commented):
left some quick comments before bed, apologies if i missed the point

bencord0 committed:
Split the test helpers into their own files

Create two new test modules, `mod common`, which will hold adapter function
to hold common patterns to call `lal::x()` functions used by the test cases.

The test cases themselves go into `mod cases`, and will be split up to
cover high-level features and their typical usage.

With rust tests, adding other files to `tests/` creates extra test
binaries. which we don't want. Instead, `#[test]` cases can be split
between multiple files if they are in a sub modules.

bencord0 committed:
test_fetch: fetch core and dev dependencies

This adds helpers for the basic fetch/build/publish workflow.

Building and publishing a named test component is going to be
foundational for other tests, so I've included an additional
`publish_component` test helper (which will clean up the
component once the test is finished).

bencord0 committed:
test_backend.rs

bencord0 committed:
test_build.rs

Exercise basic fetch and build routines.
Exercise building with other options and modes.

Add `build_with_options_and_modes` helper.

bencord0 committed:
test_publish.rs

bencord0 committed:
test_stash.rs

bencord0 committed:
test_update.rs

Clippy; conditional_chain -> swap chained `if` for `match`/`cmp`.

Add tests for lal::update. These tests document the current
behaviour of the command, but I'm not sure if they are the
behaviours that we want this command to have. As noted in
the comments, some of these behaviours are inconsistent.

e.g. not saving when --save is passed. or not installing a
devDependency when --dev is passed.

Oddly, it is also possible to depend on the same component,
but at different versions (one in `dependencies` and the
other in `devDependencies`).

bencord0 committed:
test_status.rs

This commit introduces a `stash_component` helper.
Similar to `publish_component`, except that this takes a
non-integer string and cannot be published.

Note: when deciding to use publish or stash,
you can stash a release build, but you can't publish a non-release build.

bencord0 committed:
test_shell.rs

Commit the `.lal/scripts/subroutine` script, instead
of generating it as part of the test.

bencord0 committed:
test_verify.rs

bencord0 committed:
test_list.rs

bencord0 committed:
test_remove.rs

Remove an unnecessary call to `to_string()`.

bencord0 committed:
test_export.rs

bencord0 committed:
test_query.rs

bencord0 committed:
test_clean.rs

bencord0 committed:
test_init.rs

bencord0 committed:
test_envs.rs

bencord0 committed:
test_propagate.rs

Remove an unneeded `.clone()`.
Add "default" to prop-leaf's supported environments.
It's dependencies don't have to be in default too.

bencord0 committed:
test_upgrade.rs

The self-upgrader tests are marked as ignored; which is unfortunate,
since they would have caught a bug.

In the meantime, I can test these code paths with the following patch.

---

diff --git a/src/upgrade.rs b/src/upgrade.rs
index 1b31a7ce..d91498ed 100644
--- a/src/upgrade.rs
+++ b/src/upgrade.rs
@@ -147,7 +147,11 @@ fn upgrade_exe(latest: &LatestLal, exe: &ExeInfo) -> LalResult<()> {
 /// If a newer version found (> in semver), and it's a static executable,
 /// then an executable upgrade is attempted from the new release url.
 pub fn upgrade(silent: bool) -> LalResult<bool> {
-    let latest = get_latest_lal_version()?;
+    // let latest = get_latest_lal_version()?;
+    let latest = LatestLal {
+        url: "http://localhost/".to_string(),
+        version: semver::Version::new(0, 0, 0),
+    };
     let exe = identify_exe()?;

     if latest.version > exe.version {

bencord0 committed:
run clippy (stable) 1.40 checks

bencord0 commented:
There are quite a few commits in this PR now.

Each commit is responsible for lifting out each of `lal`'s main commands' tests, and can be reviewed on a commit-by-commit basis.

The whole series should also be individually testable, and you can loop through the commits and check the commits with a command like ...

```
git rebase -i origin/master --exec "cargo +nightly fmt -- --files-with-diff --check" --exec "cargo test --no-default-features"
```

or, if you have time to grab a coffee, run the slower tests under docker.
```
git rebase -i origin/master --exec "cargo test"
```

bencord0 commented on tests/common/build.rs:
```diff
@@ -47,3 +47,23 @@ pub fn build_with_options(
+    env_name: &str,
+    home: &Path,
+    build_opts: &lal::BuildOptions,
+    modes: lal::ShellModes,
```
I'm not sure what to do about this one.

I feel that, for consistency, this should be `&lal::ShellModes`.

But the change required to do this is
```
--- a/src/build.rs
+++ b/src/build.rs
@@ -73,9 +73,9 @@ pub fn build(
     manifest: &Manifest,
     opts: &BuildOptions,
     envname: String,
-    _modes: ShellModes,
+    modes: &ShellModes,
 ) -> LalResult<()> {
-    let mut modes = _modes;
+    let mut modes = modes.clone();
 
     // have a better warning on first file-io operation
     // if nfs mounts and stuff cause issues this usually catches it
```

Which is a bit more invasive than the scope of this PR.
I might have to queue this up in a different PR.

bencord0 commented on tests/cases/test_update.rs:
```diff
@@ -0,0 +1,719 @@
+fn assert_manifest_dev(component_dir: &Path, name: &str, version: u32) {
+    let manifest = lal::Manifest::read(&component_dir).expect("read manifest");
+    assert_eq!(
+        manifest.devDependencies.get_key_value(name),
```
This was stabilized in 1.40, please upgrade your toolchains.

https://doc.rust-lang.org/std/collections/struct.BTreeMap.html#method.get_key_value

bencord0 commented on tests/common/mod.rs:
```diff
@@ -109,3 +110,23 @@ pub fn publish_component(
+
+    match component_dirs.last() {
+        Some(last) => Ok(last.to_path_buf()),
+        None => Err(lal::CliError::UploadFailure(format!(
```
Tempted to rename this as `PublishFailure`. :thinking:

bencord0 commented on tests/cases/test_status.rs:
```diff
@@ -0,0 +1,114 @@
+    let r = status::status(&component_dir);
+    assert!(r.is_ok(), "helloworld deps not fetched");
+    let lockfile = lal::Lockfile::from_path(&component_dir.join("INPUT/heylib/lockfile.json"), "heylib");
+    assert_eq!(
```
I think I can do better than this with a `match` expression.

```
--- a/tests/cases/test_status.rs
+++ b/tests/cases/test_status.rs
@@ -1,5 +1,14 @@
 use common::*;
 use parameterized_macro::parameterized;
+use std::path::PathBuf;
+
+fn assert_missing_lockfile(path: &PathBuf, name: &str) {
+    match lal::Lockfile::from_path(path, &name) {
+        Err(lal::CliError::MissingLockfile(_)) => Ok(()),
+        Ok(_) => Err("lockfile exists".to_string()),
+        Err(ref e) => Err(format!("{}", e)),
+    }.expect("dependency not in INPUT");
+}
 
 #[parameterized(env_name = {"default", "alpine"})]
 fn test_status_without_deps(env_name: &str) {
```

bencord0 commented on tests/cases/test_remove.rs:
```diff
@@ -0,0 +1,73 @@
+    let manifest = lal::Manifest::read(&component_dir).expect("read manifest");
+    assert_eq!(manifest.dependencies.len(), 1);
+    assert_eq!(
+        manifest.dependencies.get_key_value("heylib"),
```
Needs stable rust 1.40.

https://doc.rust-lang.org/std/collections/struct.BTreeMap.html#method.get_key_value

clux replied:
that's fine, it's out :-)

bencord0 commented on tests/cases/test_propagate.rs:
```diff
@@ -0,0 +1,81 @@
+use parameterized_macro::parameterized;
+use std::path::PathBuf;
+
+fn publish_components(state: &TestState, env_name: &str) -> lal::LalResult<PathBuf> {
```
Surprisingly, rust is okay with this function name shadowing `common::publish_components`.

I wonder if it is because
a) this function doesn't call `common::publish_components`, and so there's no confusion to mistake it for recursion
b) rust can distinguish between function arity? (the state of https://github.com/rust-lang/rfcs/issues/1586 suggests not)
c) we can still reference the other function with the `common::` prefix aka. universal function call syntax.

The `use *` syntax is weird.

clux replied:
i thought visibility of names like this was identical to variable shadowing; always trying the most shadowed name first. Pretty sure `c)` is true here at least.

clux commented on tests/cases/test_update.rs:
```diff
@@ -0,0 +1,720 @@
+    );
+    assert!(r.is_ok(), "updated heylib=2 prop-leaf=2");
+
+    // XXX: Bug? Why isn't the core dependency updated in the manifest?
```
isn't this just because `save==false`?

clux commented:
Presumably this does a lot more linking during testing, since it's now creating one binary per test file?

At any rate, this looks much nicer than before. Just giving it a tick in the airport in case you are blocked :-)

bencord0 commented:
> more linking during testing, since it's now creating one binary per test file

Actually, no. just the single test binary is created. The tests do take longer to run, but that's because there are more (focused) tests (although the coverage of the tests is still pretty much the same, I'll add output/return value testing later on).

The rule about one-binary-per-file only applies to `.rs` files in the top-level `tests/` directory.
Extra files in `tests/common` and `tests/cases` are compiled and linked once. Looking at other examples on github, `mod common` also seems to be a convention for naming too.
